### ISSUE TITLE
Missing HTTP error code mapping for storage.create

### ISSUE THREAD

**Issue by vokac** [xrootd/xrootd#1752]
Release candidate 5.5.0-rc1 is still [not fully compliant](https://ci.cloud.cnaf.infn.it/job/wlcg-jwt-compliance-tests/job/master/525/artifact/reports/reports/latest/se-prague-xrootd/log.html) with WLCG JWT profile `storage.create:/` scope. Attempt to overwrite existing file with just `storage.create:/` scope fails with HTTP 500 while it should fail with HTTP 403.

Current error mapping in the XRootD HTTP module is not sufficient
https://github.com/xrootd/xrootd/blob/af9b7e416aed422fcfa6cba7cbc9ead04f8f2ce3/src/XrdHttp/XrdHttpReq.cc#L986-L1009
because overwriting existing file fails with
```
220730 17:20:36 818 unknown.77:26@ui1 Xrootd_Response: 0000 sending err 3018: Unable to create /wlcg/test.create; file exists
```
and 3018 (kXR_set) error code mapping is not present => default HTTP 500 is used
https://github.com/xrootd/xrootd/blob/af9b7e416aed422fcfa6cba7cbc9ead04f8f2ce3/src/XrdHttp/XrdHttpReq.cc#L983-L984

Full server side log can be found in https://vokac.web.cern.ch/vokac/tmp/xrootd-5.5.0rc1-storage-create-http-500.log

**Comment by ffurano** [comment]
Hi,

  I see that the code misses the mapping for

  kXR_ItExists,        // 3018

  given that the error is "file exists" I'll map it to 409

  The server log (thanks!) is pretty clear about this:

220730 17:20:36 818 scitokens_Access: Grant authorization based on scopes for operation=excl_create, path=/wlcg/test.create
220730 17:20:36 818 scitokens_Access: Request username wlcgprd
220730 17:20:36 818 ofs_open: ***@***.*** Unable to create /wlcg/test.create; file exists
220730 17:20:36 818 ***@***.*** Xrootd_Response: 0000 sending err 3018: Unable to create /wlcg/test.create; file exists
220730 17:20:36 818 http_Req:  XrdHttpReq::Error
220730 17:20:36 818 ***@***.*** http_Req: PostProcessHTTPReq req: 4 reqstate: 0
220730 17:20:36 818 ***@***.*** http_Req: PostProcessHTTPReq mapping Xrd error [3018] to status code [500]
220730 17:20:36 818 ***@***.*** http_Protocol: Sending resp: 500 header len:105

Fabrizio

**Comment by vokac** [comment]
HTTP 409 is not compatible with expectation of our WLCG Compliance testsuite, so this should be discussed with StoRM and dCache developers to get same behavior for failed operation with `storage.create:/` scope defined in the [WLCG JWT profile](https://github.com/WLCG-AuthZ-WG/common-jwt-profile/blob/master/profile.md#capability-based-authorization-scope).

**Comment by ffurano** [comment]
Interesting situation indeed.
@abh3 the HTTP plugin does not know "a posteriori" that the request had just the permission to create but not to overwrite. I think that the only way to have a 403 is for the xrootd bridge to return kXR_NotAuthorized instead than kXR_ItExists for this case.

The alternative to make 409 a legal return code (which would make sense to me) could be discussed anyway.

**Comment by vokac** [comment]
@bbockelm with 5.5rc3 I'm [getting HTTP 409](https://ci.cloud.cnaf.infn.it/job/wlcg-jwt-compliance-tests/job/master/553/artifact/reports/reports/latest/joint-log.html#s1-s13-s2-t14) which is not compatible with WLCG JWT compliance tests that expects 403.

**Comment by bbockelm** [comment]
@vokac - unfortunately, this patch was not merged before 5.5.0-rc3 was tagged.  It's currently in the stable-5.5.x branch; would you have the ability to use a nightly build?

**Comment by vokac** [comment]
I upgraded my testbed to the [XRootD CI build 20220824](https://storage-ci.web.cern.ch/xrootd/master/cc-7/x86_64/20220824/) and now everything looks OK

* upload new file
```bash
$ ~/apps/curl-7.71.1-el7/bin/curl --max-time 10 -s -L -i -f --show-error --capath /etc/grid-security/certificates -H "Authorization: Bearer ${BEARER_TOKEN}" --upload-file /etc/services https://xrd.farm.particle.cz:1094/wlcg/wlcg-jwt-compliance/dir2/test.new
HTTP/1.1 100 Continue
Connection: Close
Server: XrootD/v20220823.ef4f041

HTTP/1.1 200 OK
Connection: Keep-Alive
Server: XrootD/v20220823.ef4f041
Content-Length: 3

:-)
```
* attempt to overwrite file
```bash
$ ~/apps/curl-7.71.1-el7/bin/curl --max-time 10 -s -L -i -f --show-error --capath /etc/grid-security/certificates -H "Authorization: Bearer ${BEARER_TOKEN}" --oad-file /etc/services https://xrd.farm.particle.cz:1094/wlcg/wlcg-jwt-compliance/dir2/test.new
curl: (22) The requested URL returned error: 403 Forbidden
```
So, I'm happy and closing this issue...
